File: scripts/test-queues-ws.py
#!/usr/bin/env python3
"""
Test WebSocket /api/queues/ws: conectare și monitorizare mesaje 20s.
Folosire:
  export BASE_WS_URL=http://127.0.0.1:65101   # optional, default acesta
  export SESSION_TOKEN=...                    # optional; ia din browser (DevTools -> Network -> ws -> Query String sau din /api/session/token)
  python3 scripts/test-queues-ws.py

Dacă nu trimiți SESSION_TOKEN, conexiunea va primi 401 (backend cere sesiune).
"""

from __future__ import annotations

import asyncio
import json
import os
import sys
import time
from urllib.parse import urlencode, urlparse


def main() -> None:
    try:
        import websockets
    except ImportError:
        print("Lipsește pachetul: pip install websockets")
        sys.exit(1)

    base = os.environ.get("BASE_WS_URL", "http://127.0.0.1:65101").strip()
    token = os.environ.get("SESSION_TOKEN", "").strip()
    parsed = urlparse(base)
    scheme_ws = "wss" if parsed.scheme == "https" else "ws"
    netloc = parsed.netloc or parsed.path
    path = "/api/queues/ws"
    if token:
        path += "?" + urlencode({"token": token})
    uri = f"{scheme_ws}://{netloc}{path}"
    print(f"Conectare la {uri} ...")
    if not token:
        print("(Nu e setat SESSION_TOKEN; backend-ul poate returna 401.)")

    async def run() -> None:
        try:
            async with websockets.connect(
                uri,
                ping_interval=None,
                ping_timeout=None,
                close_timeout=10,
            ) as ws:
                print("Conectat. Aștept mesaje 20s (refresh la ~0s și ~15s)...\n")
                snapshot_count = 0
                start = time.monotonic()
                while (time.monotonic() - start < 20):
                    try:
                        msg = await asyncio.wait_for(ws.recv(), timeout=1.5)
                    except asyncio.TimeoutError:
                        print(".", end="", flush=True)
                        continue
                    try:
                        obj = json.loads(msg)
                        evt = obj.get("event", "?")
                        data = obj.get("data") or {}
                        if evt == "queues.snapshot":
                            snapshot_count += 1
                            queues = data.get("queues")
                            n = len(queues) if isinstance(queues, list) else 0
                            initial = data.get("initial", False)
                            err = data.get("error")
                            ts = data.get("timestamp", "")
                            print(f"\n[snapshot #{snapshot_count}] initial={initial} queues={n} error={err!r} timestamp={ts}")
                            if not initial and isinstance(queues, list) and queues:
                                for q in queues[:3]:
                                    print(f"  - {q.get('name')}: w={q.get('waiting')} a={q.get('active')} f={q.get('failed')}")
                                if len(queues) > 3:
                                    print(f"  ... și încă {len(queues) - 3} cozi")
                        else:
                            print(f"\n[{evt}] {json.dumps(data)[:200]}")
                    except json.JSONDecodeError:
                        print(f"\n[raw] {msg[:200]}")

                print(f"\n\nTotal snapshot-uri primite în 20s: {snapshot_count} (așteptat: >=2 – unul imediat, unul după ~15s)")
                if snapshot_count == 0:
                    print("Niciun snapshot – fie backend nu trimite, fie proxy-ul nu transmite WebSocket corect.")
                    sys.exit(1)
        except websockets.exceptions.InvalidStatusCode as e:
            print(f"Eroare HTTP: {e.status_code}")
            if e.status_code == 401:
                print("401 Unauthorized – setează SESSION_TOKEN (token din browser sau din /api/session/token).")
            sys.exit(1)
        except Exception as e:
            print(f"Eroare: {e}")
            sys.exit(1)

    asyncio.run(run())


if __name__ == "__main__":
    main()

File: scripts/test-queues-ws.ts
/**
 * Test WebSocket /api/queues/ws: conectare și monitorizare mesaje 20s.
 *
 * Folosire (recomandat în dev): rulează cu secretele din OpenBao – tokenul este rezolvat automat din DB + SHOPIFY_API_SECRET:
 *   scripts/with-secrets.sh pnpm tsx scripts/test-queues-ws.ts
 *
 * Opțional:
 *   BASE_WS_URL=http://127.0.0.1:65101  (default: 65101 prin proxy web-admin)
 *   SESSION_TOKEN=...                    (dacă e setat, se folosește; altfel se generează din DB + secret)
 */

import { createHmac } from 'node:crypto';
import pg from 'pg';
import WebSocket from 'ws';

interface SessionData {
  shopId: string;
  shopDomain: string;
  createdAt: number;
}

interface QueueSnapshot {
  name: string;
  waiting: number;
  active: number;
  failed: number;
}

interface SnapshotPayload {
  queues?: QueueSnapshot[];
  initial?: boolean;
  error?: string;
  timestamp?: string;
}

interface WsMessage {
  event: string;
  data: SnapshotPayload;
}

const base = (process.env['BASE_WS_URL'] ?? 'http://127.0.0.1:65101').trim();
const wsScheme = base.startsWith('https') ? 'wss' : 'ws';
const u = new URL(base);

function createSessionToken(data: SessionData, secret: string): string {
  const payload = Buffer.from(JSON.stringify(data)).toString('base64url');
  const signature = createHmac('sha256', secret).update(payload).digest('base64url');
  return `${payload}.${signature}`;
}

async function resolveSessionToken(): Promise<string> {
  const explicit = (process.env['SESSION_TOKEN'] ?? '').trim();
  if (explicit) {
    return explicit;
  }

  const databaseUrl = process.env['DATABASE_URL'] ?? process.env['MIGRATION_DATABASE_URL'];
  const secret = (process.env['SHOPIFY_API_SECRET'] ?? '').trim();
  if (!databaseUrl || !secret) {
    console.error('Eroare: SESSION_TOKEN ne setat și (DATABASE_URL + SHOPIFY_API_SECRET) lipsesc.');
    console.error(
      'Rulează cu secretele dev: scripts/with-secrets.sh pnpm tsx scripts/test-queues-ws.ts'
    );
    process.exit(1);
  }

  const pool = new pg.Pool({ connectionString: databaseUrl });
  try {
    const res = await pool.query<{ id: string; shopify_domain: string }>(
      'SELECT id, shopify_domain FROM shops ORDER BY created_at LIMIT 1'
    );
    const row = res.rows[0];
    if (!row) {
      console.error('Eroare: nu există niciun shop în DB. Conectează mai întâi magazinul Shopify.');
      process.exit(1);
    }
    const token = createSessionToken(
      { shopId: row.id, shopDomain: row.shopify_domain, createdAt: Date.now() },
      secret
    );
    console.info('Token sesiune generat din DB (shop:', row.shopify_domain, ')');
    return token;
  } finally {
    await pool.end();
  }
}

async function main(): Promise<void> {
  const token = await resolveSessionToken();
  const path = `/api/queues/ws?token=${encodeURIComponent(token)}`;
  const uri = `${wsScheme}://${u.host}${path}`;

  console.info('Conectare la', uri, '...');

  const ws = new WebSocket(uri);
  let snapshotCount = 0;
  const start = Date.now();
  const timeout = 20_000;

  ws.on('open', () => {
    console.info('Conectat. Aștept mesaje 20s (refresh la ~0s și ~15s)...\n');
  });

  ws.on('message', (raw: WebSocket.RawData) => {
    const text = Buffer.isBuffer(raw)
      ? raw.toString('utf8')
      : raw instanceof ArrayBuffer
        ? Buffer.from(raw).toString('utf8')
        : Array.isArray(raw)
          ? Buffer.concat(raw).toString('utf8')
          : '';
    try {
      const obj = JSON.parse(text) as WsMessage;
      const evt = obj.event ?? '?';
      const payload = obj.data;
      if (evt === 'queues.snapshot') {
        snapshotCount++;
        const queues = payload.queues;
        const n = Array.isArray(queues) ? queues.length : 0;
        const initial = payload.initial === true;
        const err = payload.error;
        const ts = payload.timestamp ?? '';
        console.info(
          `\n[snapshot #${String(snapshotCount)}] initial=${String(initial)} queues=${String(n)} error=${JSON.stringify(err)} timestamp=${ts}`
        );
        if (!initial && Array.isArray(queues) && queues.length > 0) {
          for (const q of queues.slice(0, 3)) {
            console.info(
              `  - ${q.name}: w=${String(q.waiting)} a=${String(q.active)} f=${String(q.failed)}`
            );
          }
          if (queues.length > 3) {
            console.info(`  ... și încă ${String(queues.length - 3)} cozi`);
          }
        }
      } else {
        console.info(`\n[${evt}]`, JSON.stringify(payload).slice(0, 200));
      }
    } catch {
      console.info('\n[raw]', text.slice(0, 200));
    }
  });

  ws.on('close', (code: number, reason: Buffer) => {
    const elapsed = ((Date.now() - start) / 1000).toFixed(1);
    console.info(
      `\n\nConexiune închisă: code=${String(code)} reason=${String(reason) || '(none)'} după ${elapsed}s.`
    );
    console.info(
      `Total snapshot-uri primite: ${String(snapshotCount)} (așteptat: >=2 – imediat + după ~15s)`
    );
    if (code === 1006) {
      console.info('(Code 1006 = conexiunea închisă anormal, de ex. 401 Unauthorized la upgrade.)');
    }
    if (snapshotCount === 0 && code !== 1000) {
      console.error(
        'Niciun snapshot – fie 401 (setează SESSION_TOKEN), fie proxy-ul nu transmite WebSocket (verifică ws: true în vite.config.ts și repornește web-admin).'
      );
      process.exit(1);
    }
  });

  ws.on('error', (err: Error) => {
    console.error('Eroare WebSocket:', err.message);
    process.exit(1);
  });

  await new Promise<void>((resolve) => {
    setTimeout(resolve, timeout);
  });
  ws.close();
}

main().catch((err: unknown) => {
  console.error(err);
  process.exit(1);
});
